File: desktop/src/protocol/agents.ts
import type { ConfigOption } from "./runtime";

// ── Agent registry ──────────────────────────────────────────────────────────

export interface AgentConfig {
  id: string;
  displayName: string;
  acpCommand: string;
  enabled: boolean;
  /** Cached model/effort selectors from the agent's last ACP probe. */
  models: ConfigOption[];
  /** Last model the user explicitly picked; used as default for new tasks. */
  lastModel?: string;
}

export interface DetectedAgent {
  id: string;
  displayName: string;
  installed: boolean;
  defaultAcpCommand: string;
  installHint: string;
  version?: string;
  latestVersion?: string;
  /** "current" | "behind" | "missing" | "unknown" */
  status: string;
  installCommand?: string;
  updateCommand?: string;
  canManage: boolean;
}

/**
 * One registered login for an agent. Never carries credentials — the daemon
 * keeps those in its vault and only reports what the switcher shows.
 */
export interface AccountInfo {
  /** Stable id, "<agent>:<slug>". */
  id: string;
  agentId: string;
  label: string;
  email?: string;
  /** Plan or seat tier, when the agent reports one. */
  plan?: string;
  /** Whether new sessions for this agent use this account. */
  active: boolean;
}

// ── Agent rate limits ───────────────────────────────────────────────────────

/** One usage window (session, weekly, …) on an agent account. */
export interface AgentLimitWindow {
  /** "five_hour" | "seven_day" | "seven_day_opus" | "seven_day_sonnet" | "primary" | "secondary"
   *  — plus the synthetic "rate_limited" (usedPercent 100, resetsAt from the
   *  server's Retry-After) when the harness has hard-stopped the account. */
  id: string;
  /** Human label, e.g. "Session" | "Weekly" | "Weekly (Opus)". */
  label: string;
  /** 0..100 */
  usedPercent: number;
  /** Unix SECONDS when the window resets. */
  resetsAt?: number;
  windowMinutes?: number;
}

/** Rate-limit state for one registered agent account. */
export interface AgentAccountLimits {
  /** "<agent>:<slug>", e.g. "claude:personal" */
  accountId: string;
  /** "claude" | "codex" | "opencode" */
  agentId: string;
  label: string;
  /** Currently-active account for that agent. */
  active: boolean;
  /** "plus", "max20", … */
  plan?: string;
  windows: AgentLimitWindow[];
  /** Any window is at 100%. */
  exhausted: boolean;
  /** Unix SECONDS when this snapshot was taken. */
  fetchedAt: number;
  /** "api" | "local" | "unknown" */
  source: string;
  /** Per-account failure, e.g. "not logged in". */
  error?: string;
}

/** Response of the `listAgentLimits` RPC. */
export interface ListAgentLimitsResult {
  accounts: AgentAccountLimits[];
}

/**
 * API-equivalent spend for one harness: what the usage *would* cost at API
 * rates, never an amount charged to the user. A Max/Plus/Team subscription is
 * billed nothing of the sort. Reported per harness, not per account — the cost
 * stream carries no account id.
 */
export interface AgentSpend {
  /** "claude" | "codex" | "opencode" */
  agentId: string;
  /** Accrued in the last 24h; null when unknown. */
  todayUsd: number | null;
  /** All-time total we can account for; null when unknown. */
  totalUsd: number | null;
  /** Tasks that contributed spend. */
  tasks: number;
  /** False when this harness never reports cost at all. */
  reported: boolean;
}

/** Response of the `listAgentSpend` RPC. */
export interface ListAgentSpendResult {
  agents: AgentSpend[];
}

export interface LspStartResult {
  serverId: string;
  available: boolean;
  rootPath: string;
}

/** Install state of one editor language's language server (mirrors Rust). */
export interface DetectedLanguageServer {
  id: string;
  language: string;
  installed: boolean;
  version?: string | null;
  latestVersion?: string | null;
  /** "current" | "behind" | "missing" | "unknown" */
  status: string;
  installCommand?: string | null;
  updateCommand?: string | null;
  canManage: boolean;
  installHint: string;
}

File: crates/warpforge-protocol/src/agents.rs
//! Coding agents and the accounts they run under, plus language servers.

use crate::ConfigOption;
use serde::{Deserialize, Serialize};

/// A user-configured ACP agent (persisted in SQLite, managed via UI).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AgentConfig {
    pub id: String,
    pub display_name: String,
    /// The ACP server command run as `sh -c <acp_command>`.
    pub acp_command: String,
    pub enabled: bool,
    /// Cached model/effort selectors the agent exposed via its last ACP
    /// `session/update` (`configOptions`). Probed once on enable and refreshed
    /// on daemon startup so the New Task view can offer a model picker before
    /// any prompt is sent. Empty when the probe failed or the agent exposes no
    /// model selector.
    #[serde(default)]
    pub models: Vec<ConfigOption>,
    /// Last model the user explicitly picked when starting a task with this
    /// agent. Used as the default for new tasks and for orchestrator-spawned
    /// sub-agents (which have no UI to pick from). `None` until the first
    /// explicit choice.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_model: Option<String>,
}

/// An agent candidate surfaced by auto-detection (sent in the setup popup).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DetectedAgent {
    pub id: String,
    pub display_name: String,
    pub installed: bool,
    pub default_acp_command: String,
    pub install_hint: String,
    /// Installed version, when it could be determined.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    /// Latest published version (from the npm registry), when known.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latest_version: Option<String>,
    /// Freshness verdict: "current" | "behind" | "missing" | "unknown".
    pub status: String,
    /// Shell command that installs the agent (npm/brew). None when there is no
    /// automatable install (unknown package manager).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub install_command: Option<String>,
    /// Shell command that updates the agent to latest, derived from how the
    /// existing binary was installed. None when we can't update it safely.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub update_command: Option<String>,
    /// Whether the daemon can run an automated install/update for this agent.
    pub can_manage: bool,
}

/// One registered login for an agent. Carries only what the switcher shows —
/// credentials never cross this boundary.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AccountInfo {
    /// Stable id, `"<agent>:<slug>"`.
    pub id: String,
    pub agent_id: String,
    /// User-facing name ("personal", "work"). Editable.
    pub label: String,
    /// Account email, read out of the agent's own credential metadata.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    /// Plan or seat tier, when the agent reports one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan: Option<String>,
    /// Whether new sessions for this agent use this account.
    pub active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AgentLimitWindow {
    pub id: String,
    pub label: String,
    pub used_percent: f64,
    pub resets_at: Option<i64>,
    pub window_minutes: Option<u64>,
}

/// API-equivalent spend reported by the harness (what the usage *would* cost
/// at API rates), not an amount charged to the user. On a Max/Plus/Team
/// subscription nothing of the sort is billed.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentSpend {
    pub agent_id: String,
    pub today_usd: Option<f64>,
    pub total_usd: Option<f64>,
    pub tasks: u32,
    pub reported: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AgentAccountLimits {
    pub account_id: String,
    pub agent_id: String,
    pub label: String,
    pub active: bool,
    pub plan: Option<String>,
    pub windows: Vec<AgentLimitWindow>,
    pub exhausted: bool,
    pub fetched_at: i64,
    pub source: String,
    pub error: Option<String>,
}

/// One supported editor language and its language-server install state, sent
/// by [`Method::LanguageServersDetect`].
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DetectedLanguageServer {
    /// Editor language id, matching the daemon's server_command table
    /// (`typescript`, `rust`, `go`, `python`, `json`, `css`, `html`, `yaml`).
    pub id: String,
    /// User-facing label ("TypeScript / JavaScript", "Rust", …).
    pub language: String,
    pub installed: bool,
    /// Installed version of the server binary, when determinable.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    /// Latest published version (from the npm registry), when known.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latest_version: Option<String>,
    /// Freshness verdict: "current" | "behind" | "missing" | "unknown".
    pub status: String,
    /// Shell command that installs the server. None when there is no automatable
    /// install (unknown package manager / system-only package).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub install_command: Option<String>,
    /// Shell command that updates an installed server to latest, derived from
    /// how the existing binary was installed. None when it can't be updated.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub update_command: Option<String>,
    /// Whether the daemon can run an automated install/update for this server.
    pub can_manage: bool,
    /// Human-readable install hint shown when the server is missing.
    pub install_hint: String,
}

/// Reply to [`Method::LspStart`].
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct LspStartResult {
    pub server_id: String,
    pub available: bool,
    /// Absolute workspace root the server was rooted at. Clients build
    /// `file://` document URIs from it. Empty when unavailable.
    pub root_path: String,
}
